cmd/registry/cmd/delete/delete.go

```diff
@@ -30,145 +31,332 @@ import (
+		} else if spec, err := names.ParseSpec(name); err == nil {
+			return core.ListSpecs(ctx, client, spec, filter, false, h.apiSpecHandler())
+		} else if rev, err := names.ParseSpecRevision(name); err == nil {
+			return core.ListSpecRevisions(ctx, client, rev, filter, false, h.apiSpecRevisionHandler())
+		} else if artifact, err := names.ParseArtifact(name); err == nil {
+			return core.ListArtifacts(ctx, client, artifact, filter, false, h.artifactHandler())
+		}
+		return fmt.Errorf("unsupported pattern %+v", name)
 	}
-}
 
-func matchAndHandleDeleteCmd(
-	ctx context.Context,
-	client connection.RegistryClient,
-	taskQueue chan<- core.Task,
-	name string,
-	filter string,
-) error {
-	if api, err := names.ParseApi(name); err == nil {
-		return deleteAPIs(ctx, client, api, filter, taskQueue)
+	// If we get here, name designates an individual resource to be displayed.
+	// So if a filter was specified, that's an error.
+	if filter != "" {
+		return errors.New("--filter must not be specified for a non-collection resource")
+	}
+
+	if project, err := names.ParseProject(name); err == nil {
+		return core.GetProject(ctx, adminClient, project, h.projectHandler())
+	} else if api, err := names.ParseApi(name); err == nil {
+		return core.GetAPI(ctx, client, api, h.apiHandler())
+	} else if deployment, err := names.ParseDeployment(name); err == nil {
+		return core.GetDeployment(ctx, client, deployment, h.apiDeploymentHandler())
+	} else if deployment, err := names.ParseDeploymentRevision(name); err == nil {
+		return core.GetDeploymentRevision(ctx, client, deployment, h.apiDeploymentRevisionHandler())
 	} else if version, err := names.ParseVersion(name); err == nil {
-		return deleteVersions(ctx, client, version, filter, taskQueue)
+		return core.GetVersion(ctx, client, version, h.apiVersionHandler())
 	} else if spec, err := names.ParseSpec(name); err == nil {
-		return deleteSpecs(ctx, client, spec, filter, taskQueue)
+		return core.GetSpec(ctx, client, spec, false, h.apiSpecHandler())
+	} else if spec, err := names.ParseSpecRevision(name); err == nil {
+		return core.GetSpecRevision(ctx, client, spec, false, h.apiSpecRevisionHandler())
 	} else if artifact, err := names.ParseArtifact(name); err == nil {
-		return deleteArtifacts(ctx, client, artifact, filter, taskQueue)
+		return core.GetArtifact(ctx, client, artifact, false, h.artifactHandler())
 	} else {
-		return fmt.Errorf("unsupported resource name: see the 'registry rpc delete-' subcommands for alternatives")
+		return fmt.Errorf("unsupported pattern %+v", name)
 	}
 }
 
-func deleteAPIs(
-	ctx context.Context,
-	client *gapic.RegistryClient,
-	api names.Api,
-	filterFlag string,
-	taskQueue chan<- core.Task) error {
-	return core.ListAPIs(ctx, client, api, filterFlag, func(api *rpc.Api) error {
-		taskQueue <- &deleteTask{
-			client:       client,
-			resourceName: api.Name,
-			resourceKind: "api",
-		}
+func (h *deleteHandler) enqueue(task core.Task) {
```

Comment:
This kind of handler structure that can handle a list of any type is also a place that seems ripe for a refactored DRY solution. (I did it one way in lint.checker, you've done it slightly different.)